### PR TITLE
Cut holes into frames behind tiled clients

### PR DESCRIPTION
**Pull request by t-wissmann**
For semi-transparent clients, it is unpleasant to see the frame
decoration shining through. Hence, we now cut holes into the frame's
decoration at the area where clients are. If no true transparency is
activated, no difference should be noticeable, and for true transparency,
one sees the wallpaper or desktop window shining through.

Internally, the last 'outer' geometry of every client is used, and so
this also interacts nicely with pseudotiled windows and the grid layout.

This should improve the cosmetics and implicitly also solve #1183.

### CODE REVIEW

**Comment by t-wissmann** [comment]
For illustration, see the following screenshot and notice that one sees the blue frame border everywhere, but the red frame background does not shine through the semi-transparent client decorations:
no-frame-behind-clients

**Comment by t-wissmann** [comment]
For illustration, this is essentially the same layout before the change in this PR:
before-patch

**Comment by blarz** [comment]
Change looks good, the screenshots look promising as well. Can we add a test (based on screenshots)?

File: src/layout.h [hunk 1]
```diff
@@ -137,6 +138,7 @@ class FrameLeaf : public Frame, public FrameDataLeaf {
     DynAttribute_<int> selectionAttr_;
     DynAttribute_<LayoutAlgorithm> algorithmAttr_;
 private:
+    friend class FrameDecoration;
```
Comment (blarz) [inline src/layout.h:141]: Why is this needed?

File: src/x11-utils.h [hunk 2]
```diff
@@ -2,12 +2,14 @@
...
+#include <vector>
 
+#include "rectangle.h"
 #include "x11-types.h"
 
-// cut a rect out of the window, s.t. the window has geometry rect and a frame
-// of width framewidth remains
-void window_cut_rect_hole(Window win, int width, int height, int framewidth);
+// cut the specified rectangles out of the window
+void window_cut_rect_hole(Window win, int width, int height,
```
Comment (blarz) [inline src/x11-utils.h:11]: Should we update the name to window_cut_rect_holes?

File: src/layout.h (hunk 1 above)
Comment (blarz) [reply src/layout.h:141]: Ah nevermind, tried it out and saw the compiler error

File: src/x11-utils.h (hunk 2 above)
Comment (t-wissmann) [reply src/x11-utils.h:11]: I don't care too much :)

File: src/x11-utils.cpp [hunk 3]
```diff
@@ -6,18 +6,19 @@
 
 #include "globals.h"
 
+using std::vector;
+
 /**
  * \brief   cut a rect out of the window, s.t. the window has geometry rect and
```
Comment (blarz) [inline src/x11-utils.cpp:12]: This comment is here and in the .h file and needs to be updated, too

**Comment by t-wissmann** [comment]
> Change looks good, the screenshots look promising as well. Can we add a test (based on screenshots)?

Hm actually it should work. I didn't initially because these screenshots-testcases are quite fiddly most of the time. Also, I simply don't know what I get if I take a screenshot from a shape window, but maybe we can add test cases that ensure that we don't cut away too much.

File: src/x11-utils.h (hunk 2 above)
Comment (t-wissmann) [reply src/x11-utils.h]: I've renamed it while  updating the comment.

File: src/x11-utils.cpp (hunk 3 above)
Comment (t-wissmann) [reply src/x11-utils.cpp]: [x] fixed.

Comment (blarz) [reply src/x11-utils.cpp]: Thanks!

File: src/x11-utils.h (hunk 2 above)
Comment (blarz) [reply src/x11-utils.h]: Thank you :)

**Comment by blarz** [comment]
> > Change looks good, the screenshots look promising as well. Can we add a test (based on screenshots)?
> 
> Hm actually it should work. I didn't initially because these screenshots-testcases are quite fiddly most of the time. Also, I simply don't know what I get if I take a screenshot from a shape window, but maybe we can add test cases that ensure that we don't cut away too much.

I always appreciate more tests if possible. We can postpone it for now and add them in a separate PR and submit this one now, or we'll keep it open a little longer. Your choice. I'd also approve it in the current state.

**Review by blarz — APPROVED** [review]
